core/src/test/java/org/infinispan/test/AbstractInfinispanTest.java

```diff
@@ -72,6 +113,191 @@
+   }
+
+   private static void methodsToBeRemoved(BlockHound.Builder builder) {
+      // Total Order is to be removed!
+      builder.allowBlockingCallsInside(StateConsumerImpl.class.getName(), "awaitTotalOrderTransactions");
+
+      // The internal map only supports local mode - we need to replace with Caffeine
+      builder.allowBlockingCallsInside(RecoveryManagerImpl.class.getName(), "registerInDoubtTransaction");
+   }
+
+   private static void jgroupsMethodsAllowedToBlock(BlockHound.Builder builder) {
+      // UDP calls PlainDatagramSocketImpl#send which can block if OS buffer is full - but we just ignore
+      builder.allowBlockingCallsInside(UDP.class.getName(), "_send");
+      builder.allowBlockingCallsInside(FlowControl.class.getName(), "adjustCredit");
+      builder.allowBlockingCallsInside(Table.class.getName(), "add");
+      builder.allowBlockingCallsInside(Table.class.getName(), "get");
+
+      // THIS IS A VERY BAD HACK - because you can't instrument a static block of the class in jgroups Version class
+      // which loads a properties file
+      builder.allowBlockingCallsInside(Properties.class.getName(), "load");
+
+      // The NonBlockingCreditMap being non blocking is a bit questionable as they hold the lock while doing a lot of
+      // work - including acquiring other locks and possibly calling await
+      builder.allowBlockingCallsInside(NonBlockingCreditMap.class.getName(), "decrement");
+      builder.allowBlockingCallsInside(NonBlockingCreditMap.class.getName(), "replenish");
+   }
+
+   private static void questionableMethodsAllowedToBlock(BlockHound.Builder builder) {
+      // Component registry has a lock to protect its state - is short lived lock
+      builder.allowBlockingCallsInside(BasicComponentRegistryImpl.class.getName(), "prepareWrapperChange");
+
+      // This one should probably not be allowed - it is waiting for another component to start
+      // TODO: This might actually be a bug in the rewiring logic. (shows in StateTransferOverwritingValueTest)
+      builder.allowBlockingCallsInside(BasicComponentRegistryImpl.class.getName(), "awaitWrapperState");
+
+      // Believe these can technically block for TCP when flushing to socket
+      builder.allowBlockingCallsInside(JGroupsTransport.class.getName(), "sendCommand");
+      builder.allowBlockingCallsInside(JGroupsTransport.class.getName(), "sendResponse");
+
+      // This method calls initCacheStatusIfAbsent which can invoke readScopedState which reads scope from a file that
+      // can block the current thread while doing I/O
+      builder.allowBlockingCallsInside(ClusterTopologyManagerImpl.class.getName(), "prepareJoin");
+      builder.allowBlockingCallsInside(ClusterTopologyManagerImpl.class.getName(), "updateClusterState");
+   }
+
+   private static void allowTestsToBlock(BlockHound.Builder builder) {
+      builder.allowBlockingCallsInside(EvictionWithConcurrentOperationsTest.class.getName() + "$Latch", "blockIfNeeded");
```

Comment:
This doesn't look very scalable... 

Since you've had to debug all of these, I wonder how much work it would be to move the tests to either return a `CompletionStage` that's completed from the main thread or to wrap all the calls in something like `TestingUtil.pretendItsNotBlocking(() -> blockingMethodCall())`. Not in this PR, of course :)

Reply:
This is what the JREBlocking class was helping with. But I hadn't taken it as far as making it as abstract as a Callable or Runnable yet, but we could very easily do that.